[IDE] A couple of fixes for `@abi`

hamishknight opened the pull request:
- Avoid tracking the `@abi` declaration in SourceKit's `AnnotatingPrinter`
- Ignore the `@abi` declaration when mangling USRs for semantic functionality since we need to be able to perform name lookup using the mangled name to retrieve the original decl. It also seems like it makes more sense in general for USRs to only be concerned about the API of a given declaration since for semantic functionality we don't care about ABI.

Resolves #85030

hamishknight commented:
@swift-ci please test

beccadax reviewed (approved):
I can't speak for the larger picture of this fix, but the logic looks sound.

bnbarham reviewed (approved):
Thanks 🙇

hamishknight committed:
[AST] Rename `mangleDeclAsUSR` -> `mangleDeclWithPrefix`

This is used by other things that don't necessarily want an IDE
USR.

hamishknight committed:
[AST] Fix setting of `DWARFMangling` in ASTMangler

Turns out this constructor was never actually setting this bit.
`mangleTypeForDebugger` also sets this to `true`, so it doesn't seem
like this caused issues in practice, but let's actually set it. This
parameter seems like it could be removed at this point, but I'll
leave that for a follow-up.

hamishknight committed:
[AST] NFC: Move `getABIDecl` onto `ASTMangler`

hamishknight committed:
[SourceKit] Fix handling of `@abi` in `AnnotatingPrinter`

Make sure we avoid adding these to the entity stack entirely, which
avoids hitting the assertion that a top-level entity isn't encountered
with a non-empty stack.

hamishknight committed:
[AST] Remove `respectOriginallyDefinedIn` parameter from `mangleAnyDecl`

This was always set to `true` except for USR mangling, where we already
have it set to `false` for IDE USRs. The other clients were:

- AutoDiff, which is just using the resulting string as a dictionary
key, so don't seem to have any preference.
- The ClangImporter, which always overrides `@_originallyDefinedIn`
anyway.

hamishknight committed:
[AST] Introduce `ASTMangler::forUSR`

Rather than setting the USR-specific bits for each USR entrypoint
into the mangler just expose a convenience constructor for USR
generation that sets them.

hamishknight committed:
[IDE] Mangle USRs using API decls rather than ABI decls

For semantic functionality the API decl is the more useful thing to
mangle, and redeclaration checking ensures we can't end up with
conflicts. This ensures we're able to perform a name lookup for the
decl based on its USR without having to maintain a separate lookup 
table for ABI names.

hamishknight commented:
@swift-ci please test
